### PR TITLE
ENT-4406: Network protocol v3 - cookie

### PR DESCRIPTION
**Pull request by olehermanse**
This PR includes several improvements to the handling of network protocol version.
It also adds a v3, named "cookie", which will be used for enterprise features only. (Reporting cookies for duplicate host identification).

Merge together: 
https://github.com/cfengine/core/pull/3907
https://github.com/cfengine/enterprise/pull/557
https://github.com/cfengine/nova/pull/1537
https://github.com/cfengine/mission-portal/pull/930
https://github.com/cfengine/documentation/pull/2229

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4034/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4034/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4034/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4035/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4035/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4035/

File: libcfnet/tls_client.c [hunk 1]
```diff
@@ -35,6 +35,7 @@
 #include <tls_client.h>
 #include <tls_generic.h>
 #include <net.h>                     /* SendTransaction, ReceiveTransaction */
+#include <protocol.h>                      /* ParseProtocolVersionNetwork() */
```
Comment (vpodzime) [inline libcfnet/tls_client.c]: Looks like this should be a part of a different commit, but not big deal, of course.

File: libcfnet/tls_client.c [hunk 2]
```diff
@@ -206,14 +206,30 @@ int TLSClientIdentificationDialog(ConnectionInfo *conn_info,
...
     {
-        /* TODO parse CFE_v%d received and use that version if it's lower. */
         wanted_version = CF_PROTOCOL_LATEST;
     }
     else
     {
         wanted_version = conn_info->protocol;
     }
 
+    const ProtocolVersion received_version = ParseProtocolVersionNetwork(line);
```
Comment (vpodzime) [inline libcfnet/tls_client.c]: What populates the `conn_info->protocol` which is compared above?

File: libcfnet/tls_client.c [hunk 3]
```diff
@@ -206,14 +206,30 @@ int TLSClientIdentificationDialog(ConnectionInfo *conn_info,
...
+
+    if (received_version < wanted_version && ProtocolIsTLS(received_version))
+    {
+        // Downgrade as long as it's still TLS
+        wanted_version = received_version;
+    }
+    else if (ProtocolIsUndefined(received_version)
+             || ProtocolIsClassic(received_version))
+    {
+        Log(LOG_LEVEL_ERR, "Server sent a bad version number! (0a)");
```
Comment (vpodzime) [inline libcfnet/tls_client.c]: `(0a)`? Keeping the tradition alive? :smile:

File: cf-serverd/cf-serverd-enterprise-stubs.c [hunk 4]
```diff
@@ -58,6 +58,11 @@ ENTERPRISE_FUNC_1ARG_DEFINE_STUB(bool, ReceiveCollectCall, ARG_UNUSED ServerConn
     return false;
 }
 
+ENTERPRISE_FUNC_1ARG_DEFINE_STUB(bool, ReturnCookies, ARG_UNUSED ServerConnectionState *, conn)
```
Comment (vpodzime) [inline cf-serverd/cf-serverd-enterprise-stubs.c]: *"This command is used for hosts using duplicate identity."* is a bit misleading, I think. After all, the protocol is used for all hosts, if possible, and it's **purpose** is to identify hosts with duplicate identity. Also the changelog entry should be more explanatory, I think.

File: cf-serverd/server_tls.c [hunk 5]
```diff
@@ -1078,6 +1078,29 @@ bool BusyWithNewProtocol(EvalContext *ctx, ServerConnectionState *conn)
...
+    case PROTOCOL_COMMAND_COOKIE:
+    {
+        if (ConnectionInfoProtocolVersion(conn->conn_info) < CF_PROTOCOL_COOKIE)
+        {
+            goto protocol_error;
+        }
+        if (acl_CheckExact(query_acl, "delta",
+                           conn->ipaddr, conn->revdns,
+                           KeyPrintableHash(ConnectionInfoKey(conn->conn_info)))
+            == false)
```
Comment (vpodzime) [inline cf-serverd/server_tls.c]: I'd add a variable with some descriptive name to store the result of `acl_CheckExact` and then used that in the `if` (with `!` instead of `== false`).

File: libpromises/dbm_api.h [hunk 6]
```diff
@@ -55,6 +56,7 @@ typedef enum
     dbid_bundles,   // Deprecated
     dbid_packages_installed, //new package promise installed packages list
     dbid_packages_updates,   //new package promise list of available updates
+    dbid_cookies,
```
Comment (vpodzime) [inline libpromises/dbm_api.h]: Maybe add a comment for this one too?

File: cf-testd/cf-testd.c [hunk 7]
```diff
@@ -371,6 +371,16 @@ static bool CFTestD_BusyLoop(
     // See BusyWithNewProtocol() in server_tls.c for how to add other commands
     switch (GetCommandNew(recvbuffer))
     {
+    case PROTOCOL_COMMAND_COOKIE:
+    {
+        const char response[] = "COOKIES 0 0";
+        const size_t size = sizeof(response) - 1;
+        assert(size == strlen(response));
```
Comment (vpodzime) [inline cf-testd/cf-testd.c]: This assert is really not necessary.

File: cf-serverd/cf-serverd-enterprise-stubs.c (hunk 4 above)
Comment (olehermanse) [reply cf-serverd/cf-serverd-enterprise-stubs.c]: > ~"This command is used for hosts using duplicate identity."~ is a bit misleading

~English~ is a bit misleading.

Natural language is a bit misleading.

> "This command is used for **detecting** hosts using duplicate identity."

I see that I made a mistake, it is missing `detecting`.

> After all, the protocol is used for all hosts, if possible and it's **purpose** is to identify hosts with duplicate identity. Also the changelog entry should be more explanatory, I think.

Again, see above. I think, after adding `detecting`, it makes sense. Added.

File: libcfnet/tls_client.c (hunk 3 above)
Comment (olehermanse) [reply libcfnet/tls_client.c]: Of course!

File: libcfnet/tls_client.c (hunk 2 above)
Comment (olehermanse) [reply libcfnet/tls_client.c]: This client code is used in many different scenarios. For filecopy it is based on policy:

```
    ConnectionFlags flags = {
        .protocol_version = DecideProtocol(ctx, fc->protocol_version),
        .cache_connection = !background,
        .force_ipv4 = fc->force_ipv4,
        .trust_server = fc->trustkey,
        .off_the_record = false
    };
    [...]
    AgentConnection *conn = ServerConnection(servername, port, conntimeout, flags, &err);
```

In other places, like cf-hub it will always attempt latest:

```
    ConnectionFlags flags = { .protocol_version = CF_PROTOCOL_LATEST };
    return ServerConnection(ip_address, port, 30, flags, err_code);
```

File: libpromises/dbm_api.h (hunk 6 above)
Comment (olehermanse) [reply libpromises/dbm_api.h]: Added.

File: cf-testd/cf-testd.c (hunk 7 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Removed.

File: libcfnet/tls_client.c (hunk 1 above)
Comment (olehermanse) [reply libcfnet/tls_client.c]: Yep, commit after, thanks, fixed.

File: libcfnet/protocol.c [hunk 8]
```diff
@@ -46,18 +46,15 @@ ProtocolVersion ParseProtocolVersionNetwork(const char *const s)
 
 ProtocolVersion ParseProtocolVersionPolicy(const char *const s)
 {
-    if (s == NULL ||
-        strcmp(s, "0") == 0 ||
-        strcmp(s, "undefined") == 0)
+    if ((s == NULL) || (strcmp(s, "0") == 0) || (strcmp(s, "undefined") == 0))
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libcfnet/protocol.c]: Added.

File: cf-serverd/server_tls.c (hunk 5 above)
Comment (olehermanse) [reply cf-serverd/server_tls.c]: Added.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4039/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4039/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4039/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4046/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4046/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4046/

**Comment by olehermanse** [comment]
@cf-bottom And yet another jenkins build, please :)

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4047/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4047/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4047/

**Comment by olehermanse** [comment]
@cf-bottom jenkins again, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4050/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4050/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4050/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4051/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4051/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4051/

**Comment by olehermanse** [comment]
@cf-bottom one more time in jenkins :)

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4055/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4055/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4055/
